### ISSUE TITLE
Improvements for the v3

### ISSUE THREAD

**Issue by Elhebert** [Elhebert/laravel-sri#57]
Let's gather some ideas on how to improve the package for its next version.

**Comment by Elhebert** [comment]
### Laravel Components
The first thing that comes to my mind would be dropping the Blade directives in favors of Laravel Components
This will increase the DX and decrease the difficulty of using Blade directives that don't really support multiple arguments.

With Laravel 7 components we will be able to use 

```html
<x:sri.mix src="/public/js/app.js"/>
<x:sri.asset src="/public/js/app.js"/>
``` 

instead of doing 

```html
@mixSri("/public/js/app.js")
@assetSri("/public/js/app.js")
``` 

This will also enable us to get rid of all the HTML generation code. We could also create dedicated components for the CSS and the JavaScript tags that would be used the exact same way as the standard `link` and `script` tags would be used. it would make the transition so much easier.

Example:

```html
<x:sri.script src="/public/js/app.js />
<!-- <script src="{{ asset("/public/js/app.js") }} /> -->

<x:sri.script-mix src="/public/js/app.js />
<!-- <script src="{{ mix("/public/js/app.js") }} /> -->

<x:sri.link href="/public/css/app.css />
<!-- <link href="{{ asset("/public/css/app.css") }} rel="stylesheet" /> -->

<x:sri.link-mix href="/public/css/app.css />
<!-- <link href="{{ mix("/public/css/app.css") }} rel="stylesheet" /> -->
```

**Comment by juliomotol** [comment]
This looks great! I like the `<x:sri.script />` and `<x:sri.link />` better since the `mix()` and `asset()` are exposed on the top level. This will make the component generation much simpler.

I did mention i have a few concern on my recent PR. Its about the security and performance for assets loaded via CDNs. It seems that it is always computing for the hash of a file from a CDN. 

https://github.com/Elhebert/laravel-sri/blob/1e5e121e48a6064ba25009349b152ed0354bc518/src/Sri.php#L111-L112

#### Security

What if the CDN has been compromised? An attacker could inject any scripts and that file will always be able to load in our pages since we're always computing the hash for that file setting it in the `integrity` attribute.

#### Performance

I don't know how heavy it is to compute for the hash but I do know that the `file_get_contents()` can eat up some precious loading time. We could cache the computed hash for a file for better performance. Also doing so could help with the security concern earlier.

**Comment by Elhebert** [comment]
Something that we could try to do is to add a `sri:generate-hash` artisan command that would get the file, generate a hash for all the files and write it inside the configuration file.

Not sure if that's possible tho, but that would be very neat, as you could add that as a deployment step.

**Comment by juliomotol** [comment]
Interesting, I'll look it up if its doable.

### Update:

I've been reading Laravel's `config:cache` and `route:cache` commands and here is how they generate a cache:

- Generate a variable with all the data you want to cache.
- Save it in a `.php` file and place it inside `./boostrap/cache`.

So in our case, we can render all the blade templates and perform a regex to using this:

```
/<(?:link\shref="|script\ssrc=")((?:https|http|\/\/)[^\s"]*)".*integrity="([^"]*)">/
```

This regex will capture the `href`/`src` and the `integrity` values for `link` and `script`. See [RegExr Sandbox](https://regexr.com/58f7j)

After that, we can have that placed in a cache inside `./bootstrap/cache/sri.php`.

The only problem I'm having is how to load `./bootstrap/cache/sri.php` in our config. Maybe we could do something like

#### subresource-integrity.php

```php
...
    'hashes' => array_merge(require base_path('/bootstrap/cache/sri.php'), [
        // 'URL' => 'your_hash',
    ])

    'dangerous_allow_third_party_assets' => false,
...
```

Also, we'll add a new config `dangerous_allow_third_party_assets` where it is set to `false` by default.

In our `sri:generate-hash` command, we'll temporarily enable this config so that we could compute the hash for those assets.

**Comment by Elhebert** [comment]
If we introduce Laravel Component, we might have a "central" class that is used by every instance of script/link tags.

Maybe there's a way to be able to capture the src/href parameters from the component classes and inject them in our cache script. It's a long shot though 🤔

**Comment by juliomotol** [comment]
I guess we could do that, but for **Single Responsibility Principle**'s sake, we should introduce a new class that will handle interaction with the cache script. 

Just to get in the same page, here is whats gonna happen when you are to run the `sri:generate-hash` command:
1. Temporarily enable `dangerous_allow_third_party_assets` config.
2. Render all blade files.
3. When the renderer hits a `x:sri.*` component, The component class is able to capture the integrity hash.
4. Within the component class, we'll use the new caching class to store the integrity hash in our cache script.
    > **Note:** The caching class will constantly read and write in the cache script for each `x:sri.*` component hit. Since this command will only be ran before deployment, I think all will be well.

Now that you come to think off it, you can allow lazy caching for the hash. We could also introduce a `sri:clear-hash` command.

Also just to be inline with the other commands in Laravel, we could rename these commands:
- `sri:generate-hash` -> `sri-hash:cache`
- `sri:clear-hash` -> `sri-hash:clear`

**Comment by Elhebert** [comment]
It's been a while, but I finally found a bit of time to draft what the blade component could look like. #62.

**Comment by juliomotol** [comment]
Looks great! I'll find time this weekend to implement the security and performance concerns.

**Comment by juliomotol** [comment]
I've just hit a dead end in draft PR https://github.com/Elhebert/laravel-sri/pull/64 with the `sri:cache` command.

This was the initial plan:

> 1. Temporarily enable `dangerous_allow_third_party_assets` config.
> 2. Render all blade files.
> 3. When the renderer hits a `x:sri.*` component, The component class is able to capture the integrity hash.
> 4. Within the component class, we'll use the new caching class to store the integrity hash in our cache script.
>    > **Note:** The caching class will constantly read and write in the cache script for each `x:sri.*` component hit. Since this command will only be ran before deployment, I think all will be well.

But then I realized that blade files has variables passed on to them by the Controllers, Composers, etc. Without those payload, we wont be able to force render all the blade files.

I'm still coming up with workarounds for this. Right now what I have in mind is to just regex it out. If you have any ideas, I would appreciate it!

**Comment by Elhebert** [comment]
When I was working on the blade component I thought that we could actually use the `Cache::set()` when in production. That way we only generate the hashes on the first request 🤔

It's a "ok" alternative that would requires way less work. But I'll have to dig deeper into the cache mechanism of the blade views to see what is happening there exactly.

**Comment by juliomotol** [comment]
> When I was working on the blade component I thought that we could actually use the `Cache::set()` when in production. That way we only generate the hashes on the first request

Yeah this is currently happening. On first request, generated hash is being cached.

https://github.com/Elhebert/laravel-sri/blob/f63a5e07c2a32cab958ff444a441253fc05b4b7d/src/Sri.php#L62-L71

The only complications I'm having is with the `sri:cache` command where we are trying to cache **all** hashes for the assets.

> Something that we could try to do is to add a `sri:generate-hash` artisan command that would get the file, generate a hash for all the files and write it inside the configuration file.
> 
> Not sure if that's possible tho, but that would be very neat, as you could add that as a deployment step.

Upon checking laravel's `view:cache` command, it only compiles the blade files into php syntax, and caches the compiled php syntax. It doesn't actually tries to render the blade file. Thus it won't pass trough any logic from blade components/directives.

**Comment by Elhebert** [comment]
I had a thought for the caching:

In the `Sri.php` file we could directly use Laravel's caching system to cache what is returned by the `hash` and `html` method:

```php
// html()
return Cache::tags(['sri'])->rememberForever("html:${path}", "<value>");

// hash()
return Cache::tags(['sri'])->rememberForever("hash:${path}", "<value>");
```

Then we can still have a `sri:clear` command that would `Cache::tags(['sri'])->flush();` and only flush the `sri` related entries.

For the Blade component we can use the `@once` directive to ensure that the component is only rendering once, solving the regeneration of hashes.

This would make the whole caching way simpler IMHO.

What do you think?

**Comment by juliomotol** [comment]
I didn't knew you could tag caches. Unfornately, taggable cache is only supported on `apc`, `array`, `memcached`, and `redis` cache drivers leaving `database`, `file`, and `dynamodb` out.

Yes, it makes caching WAY simpler (I really hoped we could use it), but support for all cache drivers is out of the window.

>For the Blade component we can use the @once directive to ensure that the component is only rendering once, solving the regeneration of hashes.

Yeah, but we need to remember that some blade views can have data coming from controllers and/or view composers. In order to render all blade files, we'll need to what those data are.

I hate to say this but the `sri:cache` command is a long shot for now.

**Comment by Elhebert** [comment]
[v3.0](https://github.com/Elhebert/laravel-sri/releases/tag/3.0.0) has been tagged and released 🎉

**Comment by juliomotol** [comment]
But what about the caching feature? :rofl:

**Comment by Elhebert** [comment]
It's still there, but as you said, more work needs to be done.
I released now to avoid releasing 2 major releases (one for PHP8 support and one for these upgrades) "close" to each other.

I added in the README something about `@once` and if people add it to there code, then it'll be rendered only once. Which is good enough for now.

I'll have to dig into the caching topic however as I'm no longer working with Laravel (and no longer using the package myself) for over a year (since the new job) it's harder for me to find time for this.
